src/main/java/ui/components/pickers/MilestonePickerState.java

```diff
@@ -30,82 +34,244 @@ public MilestonePickerState(List<PickerMilestone> milestones, String userInput)
     }
 
     private void processInput(String userInput) {
+        filterMilestonesByQuerySentence(allMilestones, userInput);
+        populateBestMatchingMilestones(userInput);
         if (userInput.isEmpty()) return;
+        toggleFirstMilestone(bestMatchingMilestones);
+    }
 
-        String[] userInputWords = userInput.split(" ");
+    private void filterMilestonesByQuerySentence(List<PickerMilestone> milestones, String querySentence) {
+        String[] userInputWords = querySentence.split(" ");
         Stream.of(userInputWords)
-                .forEach(this::filterMilestones);
+                .forEach(queryWord -> filterMilestonesByQueryWord(milestones, queryWord));
+    }
 
-        toggleFirstMatchingMilestone(currentMilestonesList);
+    private void filterMilestonesByQueryWord(List<PickerMilestone> milestones, String queryWord) {
```

Comment:
For this and above, `filter` does not seem accurate. If I understand correctly, it's more of `preprocessMatchingMilestonesBy...`. By `filter`, I assume I will get something back that is filtered, whether on the variable itself or in return value. In this case, it's more of processing each milestone to set matching.